### PR TITLE
ARROW-10689: [Rust] [DataFusion] Add SQL support for CASE WHEN

### PR DESCRIPTION
**Pull request by andygrove**
This follows on from https://github.com/apache/arrow/pull/8746 and adds SQL support for CASE WHEN.

### CODE REVIEW

**Comment by andygrove** [comment]
@alamb @jorgecarleitao @Dandandan This is ready for review

File: rust/datafusion/tests/sql.rs [hunk 1]
```diff
@@ -910,6 +910,76 @@ async fn csv_query_count_one() {
...
+        vec!["1"],
+        vec!["999"],
+        vec!["999"],
+        vec!["2"],
+        vec!["999"],
+        vec!["999"],
+        vec!["999"],
+        vec!["999"],
+        vec!["999"],
+    ];
```
Comment (jorgecarleitao) [inline rust/datafusion/tests/sql.rs:944]: It is a bit difficult for me to confirm that this is correct. What do you think if we just generate a table in memory with all possibilities and test against that? IMO it is a bit more explicit and also allows to verify the cases with null values.

File: rust/datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -446,6 +446,42 @@ impl<'a, S: SchemaProvider> SqlToRel<'a, S> {
...
+                    .collect::<Result<Vec<_>>>()?;
+                let then_expr = results
+                    .iter()
+                    .map(|e| self.sql_to_rex(e, schema))
+                    .collect::<Result<Vec<_>>>()?;
+                let else_expr = if let Some(e) = else_result {
+                    Some(Box::new(self.sql_to_rex(e, schema)?))
+                } else {
+                    None
+                };
```
Comment (jorgecarleitao) [inline rust/datafusion/src/sql/planner.rs:472]: Very clean implementation! 💯

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Left an optional thing.

Thanks a lot, @andygrove , really important expression to have.

File: rust/datafusion/src/logical_plan/expr.rs [hunk 3]
```diff
@@ -805,6 +805,25 @@ fn create_name(e: &Expr, input_schema: &Schema) -> Result<String> {
             let right = create_name(right, input_schema)?;
             Ok(format!("{} {:?} {}", left, op, right))
         }
+        Expr::Case {
+            expr,
+            when_then_expr,
+            else_expr,
+        } => {
+            let mut name = String::new();
+            name = name + "CASE ";
```
Comment (Dandandan) [inline rust/datafusion/src/logical_plan/expr.rs:814]: Could be
`let mut name = "CASE  ".to_string()`

File: rust/datafusion/src/logical_plan/expr.rs [hunk 4]
```diff
@@ -805,6 +805,25 @@ fn create_name(e: &Expr, input_schema: &Schema) -> Result<String> {
...
+            if let Some(e) = expr {
+                name = name + &format!("{:?} ", e).to_string();
```
Comment (Dandandan) [inline rust/datafusion/src/logical_plan/expr.rs:816]: can use `+=` ?

**Comment by Dandandan** [comment]
LGTM. some tiny style suggestions.

**Review by Dandandan — APPROVED** [review]

**Comment by andygrove** [comment]
Thanks @Dandandan and @jorgecarleitao for the good suggestions. I have addressed those and it looks much cleaner now.

**Comment by andygrove** [comment]
I see now that there is actually a bug. When there is an ELSE condition, that value should always be returned even if the input is NULL. I will push a fix shortly.

**Pull request closed without merging**